app/models/reporter.go

```diff
@@ -0,0 +1,357 @@
+		COALESCE(MIN(Query_time_min), 0) AS query_time_min,
+		COALESCE(SUM(Query_time_sum)/SUM(total_query_count), 0) AS query_time_avg,
+		COALESCE(AVG(Query_time_med), 0) AS query_time_med,
+		COALESCE(AVG(Query_time_p95), 0) AS query_time_p95,
+		COALESCE(MAX(Query_time_max), 0) AS query_time_max
+	FROM query_global_metrics
+	WHERE instance_id = :instance_id AND start_ts BETWEEN :begin AND :end
+`
+
+const queryReportTemplate = `
+	SELECT
+		qcm.query_class_id AS query_class_id,
+		COALESCE(SUM(qcm.query_count), 0) AS query_count,
+		COALESCE(SUM(qcm.Query_time_sum), 0) AS query_time_sum,
+		COALESCE(MIN(qcm.Query_time_min), 0) AS query_time_min,
+		COALESCE(SUM(qcm.Query_time_sum)/SUM(qcm.query_count), 0) AS query_time_avg,
+		COALESCE(AVG(qcm.Query_time_med), 0) AS query_time_med,
+		COALESCE(AVG(qcm.Query_time_p95), 0) AS query_time_p95,
+		COALESCE(MAX(qcm.Query_time_max), 0) AS query_time_max,
+		qc.checksum AS checksum,
+		qc.abstract AS abstract,
+		qc.fingerprint AS fingerprint,
+		qc.first_seen AS first_seen
+	FROM query_class_metrics AS qcm
+	JOIN query_classes AS qc ON qcm.query_class_id = qc.query_class_id
+	WHERE qcm.instance_id = :instance_id AND (qcm.start_ts >= :begin AND qcm.start_ts < :end)
+		{{ if .FirstSeen }} AND qc.first_seen >= :begin {{ end }}
+		{{ if .Keyword }} AND (qc.checksum = ':keyword' OR qc.abstract LIKE '%:keyword' OR qc.fingerprint LIKE '%:keyword') {{ end }}
+	GROUP BY qcm.query_class_id
+	ORDER BY SUM(qcm.Query_time_sum) DESC
+	LIMIT :limit OFFSET :offset;
+`
+
+func (r report) Profile(instanceID uint, begin, end time.Time, rank RankBy, offset int, search string, firstSeen bool) (Profile, error) {
+	args := struct {
+		InstanceID uint `db:"instance_id"`
+		Begin      time.Time
+		End        time.Time
+		Limit      uint
+		Offset     int
+		Keyword    string
+		FirstSeen  bool
+	}{
+		InstanceID: instanceID,
+		Begin:      begin,
+		End:        end,
+		Limit:      rank.Limit,
+		Offset:     offset,
+		Keyword:    search,
+		FirstSeen:  firstSeen,
+	}
+	p := Profile{
+		// caller sets InstanceId (MySQL instance UUID)
+		Begin:  begin,
+		End:    end,
+		RankBy: rank,
+	}
+
+	intervalTime := end.Sub(begin).Seconds()
+	intervalTimeMinutes := end.Sub(begin).Minutes()
+	endTs := end.Unix()
+	amountOfPoints := int64(maxAmountOfPoints)
+	if intervalTimeMinutes < maxAmountOfPoints {
+		amountOfPoints = int64(intervalTimeMinutes)
+	}
+
+	intervalTs := int64(end.Sub(begin).Seconds()) / amountOfPoints
+
+	// get count of all rows - to calculate pagination.
+	var queryReportCountUniqueBuffer bytes.Buffer
+	if tmpl, err := template.New("queryReportCountUniqueSQL").Parse(queryReportCountUniqueTemplate); err != nil {
+		log.Fatalln(err)
+	} else if err = tmpl.Execute(&queryReportCountUniqueBuffer, args); err != nil {
+		log.Fatalln(err)
+	}
+
+	nstmtQueryReportCountUnique, err := db.PrepareNamed(queryReportCountUniqueBuffer.String())
+	defer nstmtQueryReportCountUnique.Close()
```

Comment:
should check returned error before deferring nstmtQueryReportCountUnique.Close()